ocaml/idl/markdown_backend.ml

```diff
@@ -0,0 +1,373 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+open Printf
+
+open Datamodel_types
+open Datamodel
+open Datamodel_utils
+open Dm_api
+
+open Stdext
+open Xstringext
+open Pervasiveext
+
+let escape s =
+  let sl = String.explode s in
+  let esc_char =
+    function
+    | '\\' -> "&#92;"
+    | '*' -> "&#42;"
+    | '_' -> "&#95;"
+    | '{' -> "&#123;"
+    | '}' -> "&#125;"
+    | '[' -> "&#91;"
+    | ']' -> "&#93;"
+    | '(' -> "&#40;"
+    | ')' -> "&#41;"
+    | '>' -> "&gt;"
+    | '<' -> "&lt;"
+    | '#' -> "&#35;"
+    | '+' -> "&#43;"
+    | '-' -> "&#45;"
+    | '!' -> "&#33;"
+    | c -> String.make 1 c in
+  let escaped_list = List.map esc_char sl in
+  String.concat "" escaped_list
+
+let is_prim_type = function
+  | String | Int | Float | Bool | DateTime -> true
+  | _ -> false
+
+let is_prim_opt_type = function
+  | None -> true
+  | Some (ty,_) -> is_prim_type ty
+
+let rec of_ty_verbatim = function
+  | String -> "string"
+  | Int -> "int"
+  | Float -> "float"
+  | Bool -> "bool"
+  | DateTime -> "datetime"
+  | Enum (name, things) -> name
+  | Set x -> sprintf "%s set" (of_ty_verbatim x)
+  | Map (a, b) -> sprintf "(%s -> %s) map" (of_ty_verbatim a) (of_ty_verbatim b)
+  | Ref obj -> obj ^ " ref"
+  | Record obj -> obj ^ " record"
+
+
+let rec of_ty = function
+  | String -> "string"
+  | Int -> "int"
+  | Float -> "float"
+  | Bool -> "bool"
+  | DateTime -> "datetime"
+  | Enum (name, things) -> escape name
+  | Set x -> (of_ty x) ^ " set"
+  | Map (a, b) -> "(" ^ (of_ty a) ^ " &rarr; " ^ (of_ty b) ^ ") map"
+  | Ref obj -> (escape obj) ^ " ref"
+  | Record obj -> (escape obj) ^ " record"
+
+let of_ty_opt = function
+    None -> "void" | Some(ty, _) -> of_ty ty
+
+let of_ty_opt_verbatim = function
+    None -> "void" | Some(ty, _) -> of_ty_verbatim ty
+
+let desc_of_ty_opt = function
+    None -> "" | Some(_, desc) -> desc
+
+(** Add namespaces (separated by _) to each field name *)
+let flatten stuff =
+  let rec f ns = function
+    | Field fr -> Field { fr with field_name = ns ^ fr.field_name}
+    | Namespace(ns', contents) -> Namespace("", List.map (f (ns ^ ns' ^ "_")) contents)
+  in
+  f "" stuff
+
+let string_of_qualifier = function
+  | StaticRO   -> "_RO/constructor_"
+  | DynamicRO  -> "_RO/runtime_"
+  | RW         -> "_RW_"
+
+
+let string_of_open_product release =
+  if release.internal_deprecated_since = None then "" else "**Deprecated.** "
+
+let print_field_content printer x =
```

Comment:
In my opinion it might be better to put this into the `print_field_table_of_obj` function, because right now it's quite far from where it's used, but these two functions are constructing the same table (and the order of the headers in `print_field_table_of_obj` and the order of things here have to match).